cmd/main_test.go

```diff
@@ -0,0 +1,145 @@
+	}
+	v, ok := db.apps[appID]
+	// TODO: this should bubble up as a 404
+	if !ok {
+		return nil, fmt.Errorf("no metric definition found for app %s", appID)
+	}
+	return v, nil
+}
+
+func marshalRequest(tb testing.TB, req *metrics.SendMetricRequest) io.Reader {
+	tb.Helper()
+	b, err := json.Marshal(req)
+	if err != nil {
+		tb.Fatalf("could not marshal json: %s", err.Error())
+	}
+	return bytes.NewReader(b)
+}
+
+func Test_handleMetric(t *testing.T) {
+	t.Parallel()
+	cases := []struct {
+		name       string
+		db         server.MetricsLookuper
+		body       io.Reader
+		wantStatus int
+		wantLogs   map[*slogassert.LogMessageMatch]int
+	}{
+		// TODO: more test cases if by some miracle this ugly thing doesn't get ousted in code review
+		{
+			name: "happy_single_metric",
+			db: &testMetricsDB{apps: map[string]*server.AppMetrics{"test": {
+				AppID:   "test",
+				Allowed: map[string]interface{}{"foo": struct{}{}, "bar": struct{}{}},
+			}}},
+			body: marshalRequest(t, &metrics.SendMetricRequest{
+				AppID:      "test",
+				AppVersion: "1.0",
+				Metrics:    map[string]int64{"foo": 1},
+				InstallID:  "asdf",
+			}),
+			wantStatus: 202,
+			wantLogs: map[*slogassert.LogMessageMatch]int{{
+				Message: "metric received",
+				Level:   slog.LevelInfo,
+				Attrs: map[string]any{
+					"metric.app_id":      "test",
+					"metric.app_version": "1.0",
+					"metric.name":        "foo",
+					"metric.count":       1,
+					"metric.install_id":  "asdf",
+				},
+				AllAttrsMatch: false,
+			}: 1},
+		},
+	}
+	for _, tc := range cases {
+		t.Run(tc.name, func(t *testing.T) {
+			t.Parallel()
+			ctx := context.Background()
+			h, err := renderer.New(ctx, nil,
+				renderer.WithOnError(func(err error) {
+					t.Fatalf("failed to render: %s", err.Error())
+				}))
+			if err != nil {
+				t.Fatalf("failed to setup test: %s", err.Error())
+			}
+			req := httptest.NewRequest(http.MethodPost, "/sendMetrics", tc.body)
+			req.Header.Set("User-Agent", "github.com/abcxyz/abc-updater")
+			req.Header.Set("Content-Type", "application/json")
+			req.Header.Set("Accept", "application/json")
+			logHandler := slogassert.New(t, slog.LevelInfo, nil)
+			req = req.WithContext(logging.WithLogger(req.Context(), slog.New(logHandler)))
+
+			w := httptest.NewRecorder()
+			handleMetric(h, tc.db).ServeHTTP(w, req)
+			response := w.Result()
+			defer response.Body.Close()
+
+			if got, want := response.StatusCode, tc.wantStatus; got != want {
+				t.Errorf("unexpected response code. got %d want %d", got, want)
+			}
+
+			// Normally we wouldn't test log messages, but as that is the way metrics
+			// are being exported, it seems important to do so here.
+			for k, v := range tc.wantLogs {
+				// TODO: I don't like that this panics if there are no matches, would rather handle error myself
+				// I have https://github.com/thejerf/slogassert/pull/5 to try and fix it.
+				if got, want := logHandler.AssertSomePrecise(*k), v; got != want {
```

Comment:
I don't think you need a new variable `want`, you can just rename `v` to `want`. They always have the same value.